simulation/chart.py

```diff
@@ -1,232 +1,438 @@
 
-    pen = pg.mkPen(**config)
-    return chart.plot(name=name, pen=pen)
+    return pg.mkPen(**config)
 
 
-def add_scatter_plot(
-    chart: pg.PlotItem,
-    name: str,
-    size: Optional[float] = None,
-    color=None,
-    symbol=None,
-    border=None,
-) -> pg.PlotDataItem:
-    """Adds a scatter plot to the chart.
+SymbolType = Literal["o", "x", "s"]  # dot  # cross  # square
 
-    Args:
-        name (str): Name (legend title) of the plot
-        size (float, optional): Size of the marker symbol. Defaults to None.
-        color (Any, optional): Color to fill the marker symbol. Defaults to None.
-        symbol (Any, optional): Shape of the marker symbol. Defaults to None.
-        border (Any, optional): Pen to draw border around the marker symbol. Defaults to None.
-    """
-    config = {}
 
-    if size is not None:
-        config["symbolSize"] = size
+class PenConfig(TypedDict, total=False):
+    """Configuration for pen to draw lines"""
 
-    if color is not None:
-        config["symbolBrush"] = color
+    width: float
+    """Width of the line"""
+    style: PenStyle
+    """Line style"""
+    color: Color
+    """Line color"""
 
-    if symbol is not None:
-        config["symbol"] = symbol
 
-    if border is not None:
-        config["symbolPen"] = border
+# Method to resample the data before plotting ot avoid plotting multiple line segments per pixel.
+DownsampleMethod = Literal[
+    "subsample",  # Downsample by taking the first of N samples. This method is fastest and least accurate.
+    "mean",  # Downsample by taking the mean of N samples.
+    "peak",  # Downsample by drawing a saw wave that follows the min and max of the original data. This method produces the best visual representation of the data but is slower.
+]
 
-    return chart.plot(name=name, pen=None, **config)
+DEFAULT_WIN_SIZE = (1000, 1000)
+DEFAULT_CODEC = "libx264"
+DEFAULT_PIXEL_FORMAT = "yuv420p"
 
 
-def create_video_process(
-    file_name: str,
-    size: Tuple[int, int],
-    codec: str = "libx264",
-    pixel_format: str = "yuv420p",
-) -> Popen:
-    """Creates a ffmpeg video process which accepts input from stdin.(NA in windows)
+class ChartWidget:
+    """A chart container for single group of plots and a legend.
 
     Args:
-        file_name (str): name of the file to save encoded video ouput
-        size (Tuple[int,int]): Size (width,height) of the video frame
-        codec (str, optional): Codec name to be used for encoding. Defaults to "libx264".
-        pixel_format (str, optional): Ouput pixel format. Defaults to "yuv420p".
-
-    Returns:
-        Process: Asyncronious sub process listening stdin for encoding to the file.
+        chart (PlotItem): base container plotitem
     """
-    process = (
-        ffmpeg.input(
-            "pipe:",
-            format="rawvideo",
-            pix_fmt="rgb24",
-            s=f"{size[0]}x{size[1]}",
+
+    def __init__(self, chart: pg.PlotItem) -> None:
+        self.__chart = chart
```

Comment:
From [Brett Slatkin's Effective Python 2nd Edition](https://effectivepython.com/):

> **Item 27: Prefer Public Attributes Over Private Ones**
> * Only consider using private attributes to avoid naming conflicts with subclasses that are out of your control.

In essence, you should avoid using private variables in Python, unlike in, say, Java. If you need to, you can provide a hint to downstream devs that they shouldn't touch a variable by using a single underscore, but, in general, even that is discouraged unless you have a real need to do so.

Reply:
So if this is a case for this repository, do you want me to span it to all the code in chart.py (no double underscore)

Reply:
Good question. Is the other code yours, or Alexis'?

Reply:
Only chart code changed.
https://github.com/semiotic-ai/autoagora-agents/pull/21/commits/d6582cf6a8d6916239e09a4f62d266c848672a60